File: LeetCode/Medium/0221-maximal-square/0221-maximal-square.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        ret = 0
        row_len = len(matrix)
        col_len = len(matrix[0])
        dp = [[0] * col_len for i in range(row_len)]

        for row in range(row_len):
            for col in range(col_len):
                if matrix[row][col] == "1":
                    dp[row][col] = 1
                    ret = 1

        for row in range(1, row_len):
            for col in range(1, col_len):
                if dp[row][col]:
                    dp[row][col] = min(dp[row-1][col], dp[row][col-1], dp[row-1][col-1]) + 1
                    ret = max(ret, dp[row][col])
        
        return ret ** 2

File: LeetCode/Medium/0221-maximal-square/0221-maximal-square.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int rowLen = matrix.size();
        int colLen = matrix[0].size();
        int ret = 0;
        vector<vector<int>> dp(rowLen, vector<int>(colLen));
        
        for (int i = 0; i < rowLen; ++i){
            for (int j = 0; j < colLen; ++j){
                if (matrix[i][j] == '1') {
                    dp[i][j] = 1;
                    ret = 1;
                }
            }
        }
        
        for (int i = 1; i < rowLen; ++i){
            for (int j = 1; j < colLen; ++j){
                if (dp[i][j] == 1){
                    dp[i][j] = min(dp[i-1][j], min(dp[i][j-1], dp[i-1][j-1])) + 1;
                    ret = max(ret, dp[i][j]);
                }
            }
        }
        
        return pow(ret, 2);
    }
};